renn/data/wordpiece_tokenizer_learner_lib.py

```diff
@@ -0,0 +1,398 @@
+            subtokens[length][subtoken] += count
+          else:
+            # New subtoken, add to dict.
+            subtokens[length][subtoken] = count
+        start = index
+
+    next_tokens = {}
+    # Get all tokens that have a count above the threshold.
+    for length in range(params.max_token_length, 0, -1):
+      for token, count in subtokens[length].items():
+        if count >= thresh:
+          next_tokens[token] = count
+        # Decrement the count of all prefixes.
+        if len(token) > length:  # This token includes the joiner.
+          joiner_len = len(params.joiner)
+          for i in range(1 + joiner_len, length + joiner_len):
+            prefix = token[0:i]
+            if prefix in subtokens[i - joiner_len]:
+              subtokens[i - joiner_len][prefix] -= count
+        else:
+          for i in range(1, length):
+            prefix = token[0:i]
+            if prefix in subtokens[i]:
+              subtokens[i][prefix] -= count
+
+    # Add back single-character tokens.
+    curr_tokens = ensure_all_tokens_exist(char_tokens, next_tokens,
+                                          params.include_joiner_token,
+                                          params.joiner)
+
+  vocab_words = generate_final_vocabulary(params.reserved_tokens, char_tokens,
+                                          curr_tokens)
+
+  return vocab_words
+
+
+def learn_binary_search(word_counts, lower, upper, params):
+  """Performs binary search to find wordcount frequency threshold.
+
+  Given upper and lower bounds and a list of (word, count) tuples, performs
+  binary search to find the threshold closest to producing a vocabulary
+  of size vocab_size.
+
+  Args:
+    word_counts: list of (string, int) tuples
+    lower: int, lower bound for binary search
+    upper: int, upper bound for binary search
+    params: Params namedtuple, parameters for learning
+
+  Returns:
+    list of strings, vocab that is closest to target vocab_size
+  """
+
+  thresh = (upper + lower) // 2
+  current_vocab = learn_with_thresh(word_counts, thresh, params)
+  current_vocab_size = len(current_vocab)
+
+  # Allow count to be within k% of the target count, where k is slack ratio.
+  slack_count = params.slack_ratio * params.vocab_size
+  if slack_count < 0:
+    slack_count = 0
+
+  is_within_slack = (current_vocab_size <= params.vocab_size) and (
+      params.vocab_size - current_vocab_size <= slack_count)
+
+  # We've created a vocab within our goal range (or, ran out of search space).
+  if is_within_slack or lower >= upper or thresh <= 1:
+    return current_vocab
+
+  current_vocab = None
+
+  if current_vocab_size > params.vocab_size:
+    return learn_binary_search(word_counts, thresh + 1, upper, params)
+
+  else:
+    return learn_binary_search(word_counts, lower, thresh - 1, params)
+
+
+def learn(word_counts, params):
+  """Takes in wordcounts and returns wordpiece vocabulary.
+
+  Args:
+    word_counts: list of (string, int) tuples
```

Comment:
Is word_counts assumed to be counts for sub-words or full-words? It seems what this function does is to find the right count threshold, but not splitting words into sub-units, so word_counts already contains wordpieces?

Reply:
Leaving as is for now, since this was directly forked.